Updates to ODA driver suggested by @Hallberg-NOAA Issue#277

MJHarrison-GFDL opened the pull request:
These changes follow suggestions from @Hallberg-NOAA for clarity in documentation and code readability.  These modifications should not impact existing applications (e.g. SPEAR) which are reliant on this code.

Hallberg-NOAA commented on src/ocean_data_assim/MOM_oda_driver.F90:
```diff
@@ -198,8 +202,16 @@ subroutine init_oda(Time, G, GV, US, diag_CS, CS)
-  call get_param(PF, mdl, "ASSIM_FREQUENCY", CS%assim_frequency,  &
-       "data assimilation frequency in hours")
+  call get_param(PF, mdl, "ASSIM_INTERVAL", CS%assim_interval,  &
+       "data assimilation update interval in hours",default=-1.0)
```
We are about to require that all get_param calls for real variables have a units argument.  We are also trying to isolate all unit conversion factors in the scale arguments for get param calls.  It would be better to revise the call on lines 205 and 206 to read:

` call get_param(PF, mdl, "ASSIM_INTERVAL", CS%assim_interval,  &
       "data assimilation update interval in hours", &
       units=hours, default=-1.0, scale=3600.0*US%s_to_T)`

with a similar change to the get_param call for ASSIM_FREQUENCY on the next lines, and no need for the conversion on lines 213 below.  With this minor correction, this updated code will be in alignment with the changes to the get_param optional arguments that are about to happen, and it makes the code simpler and easier to read.

Hallberg-NOAA commented on src/ocean_data_assim/MOM_oda_driver.F90:
```diff
@@ -678,8 +685,8 @@ subroutine apply_oda_tracer_increments(dt, Time_end, G, GV, tv, h, CS)
                                                     !! tendency [S T-1 -> ppt s-1]
-  real, dimension(SZI_(G),SZJ_(G),SZK_(CS%Grid)) :: T !< The updated temperature [C ~> degC]
-  real, dimension(SZI_(G),SZJ_(G),SZK_(CS%Grid)) :: S !< The updated salinity [S ~> ppt]
+  real, dimension(SZI_(G),SZJ_(G),SZK_(CS%Grid)) :: T !< The temperature tendency adjustment from DA [C T-1 ~> degC s-1]
```
The code is now correctly documented, but for clarity would it make sense to rename these local variables from `T` and `S` to `T_tend` and `S_tend`?

Hallberg-NOAA reviewed (changes requested):
Thank you very much for these updates, @MJHarrison-GFDL.  They are a significant improvement to the code!

I have a couple of other minor, specific tweaks to suggest that I think would make this even better and avoid the need for another revision to comply with an imminent change to the optional arguments for the get_param interface for real variables.

  The other changes that we might consider as a part of this would be:

1.  Change the units of the `dt` argument to  `apply_oda_tracer_increments()` from [s] to [T_to_s], 
2. Correct the units of T_bias and S_bias at about line 570  in `get_bias_correction_tracer()`, including changing the scale factor in the call to `horiz_interp_and_extrap_tracer()` for temperature from `scale=US%degC_to_C` to `scale=US%degC_to_C*US%s_to_T`, as well as the floor on T_bias at about line 592, and similarly for salinity.

MJHarrison-GFDL commented:
Thank you @Hallberg-NOAA . I have addressed your latest comments. I modified the masking in the bias_adjustment routine so that it is based on the masking flag returned from horiz_interp_and_extrap_tracer, rather than the arbitrary ceiling value used in the previous version. This change should not impact answers for MOM6_version>=20190101 .

Hallberg-NOAA commented:
Thanks for the updates to address the overall comments @MJHarrison-GFDL , but the specific comments relating to the get_param calls also need to be dealt with.  As written, this code will no longer compile once PR #300 is merged into dev/gfdl.

marshallward commented:
Gaea regression: https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/17927 :heavy_check_mark:
